jmh-benchmarks/src/main/java/org/apache/kafka/jmh/assignor/ServerSideAssignorBenchmark.java

```diff
@@ -0,0 +1,281 @@
+            for (int bucket = 0; bucket < bucketCount; bucket++) {
+                int memberStartIndex = bucket * bucketSizeMembers;
+                int memberEndIndex = Math.min((bucket + 1) * bucketSizeMembers, numberOfMembers);
+
+                int topicStartIndex = bucket * bucketSizeTopics;
+                int topicEndIndex = Math.min((bucket + 1) * bucketSizeTopics, topicCount);
+
+                Set<Uuid> bucketTopics = new HashSet<>(allTopicIds.subList(topicStartIndex, topicEndIndex));
+
+                // Assign topics to each member in the current bucket
+                for (int i = memberStartIndex; i < memberEndIndex; i++) {
+                    addMemberSpec(members, i, new HashSet<>(bucketTopics));
+                }
+            }
+        }
+
+        this.assignmentSpec = new AssignmentSpec(members);
+    }
+
+    private Optional<String> rackId(int memberIndex) {
+        return isRackAware ? Optional.of("rack" + memberIndex % NUMBER_OF_RACKS) : Optional.empty();
+    }
+
+    private void addMemberSpec(
+        Map<String, AssignmentMemberSpec> members,
+        int memberIndex,
+        Set<Uuid> subscribedTopicIds
+    ) {
+        String memberId = "member" + memberIndex;
+        Optional<String> rackId = rackId(memberIndex);
+
+        members.put(memberId, new AssignmentMemberSpec(
+            Optional.empty(),
+            rackId,
+            subscribedTopicIds,
+            Collections.emptyMap()
+        ));
+    }
+
+    private static Map<Integer, Set<String>> mkMapOfPartitionRacks(int numPartitions) {
+        Map<Integer, Set<String>> partitionRacks = new HashMap<>(numPartitions);
+        for (int i = 0; i < numPartitions; i++) {
+            partitionRacks.put(i, new HashSet<>(Arrays.asList(
+                "rack" + i % NUMBER_OF_RACKS,
+                "rack" + (i + 1) % NUMBER_OF_RACKS,
+                "rack" + (i + 2) % NUMBER_OF_RACKS
+            )));
+        }
+        return partitionRacks;
+    }
+
+    private void simulateIncrementalRebalance() {
+        GroupAssignment initialAssignment = partitionAssignor.assign(assignmentSpec, subscribedTopicDescriber);
+        Map<String, MemberAssignment> members = initialAssignment.members();
+
+        Map<String, AssignmentMemberSpec> updatedMembers = new HashMap<>();
+        members.forEach((memberId, memberAssignment) -> {
+            AssignmentMemberSpec memberSpec = assignmentSpec.members().get(memberId);
+            updatedMembers.put(memberId, new AssignmentMemberSpec(
+                memberSpec.instanceId(),
+                memberSpec.rackId(),
+                memberSpec.subscribedTopicIds(),
+                memberAssignment.targetPartitions()
+            ));
+        });
+
+        Collection<Uuid> subscribedTopicIdsForNewMember;
+        if (subscriptionModel == SubscriptionModel.HETEROGENEOUS) {
+            subscribedTopicIdsForNewMember = updatedMembers.values().iterator().next().subscribedTopicIds();
```

Comment:
Is it correct to take the subscriptions of the first member? Intuitively, I would have taken the ones from the last member (the last bucket).

Reply:
I made the buckets based on the number of members = memberCount-1 for the first assignment, so I figured we could add a member and assign it any member's subscription list.

Reply:
Since all the buckets are equally full

Reply:
Hum... Are you sure about this? I may have gotten it incorrectly but my understanding is that you use `Math.ceil`. So, if you start with 100 members, you do ceil(100 - 1 / 5) which is 20. With this, I think that the first 4 buckets will have 20 members and the last one will have 19 members in the initial assignment. I actually thought that you used `ceil` to achieve this int he first place.

Reply:
yeah, but if there are 101 members then all of them could get full and if the memberCount - 1 is <= 5
If we use the memberCount to calculate the bucket sizes first we may have some unassigned topics

Reply:
We can just assign to the last member either way for the regular cases that we put, I guess it doesn't matter if it's the first or last one when all the buckets are full